data/org.eclipse.birt.data/src/org/eclipse/birt/data/engine/impl/CancelManager.java

```diff
@@ -24,22 +26,21 @@
 
 public class CancelManager extends TimerTask {
 	//
-	private List<ICancellable> cancellableList;
+	private Map<String, ICancellable> cancellableMap = new HashMap<>();
```

Comment:
Why don't we just use a set?   If one uses a concurrent collection, would could avoid synchronize calls.  E.g, wouldn't this be better?
```
/*******************************************************************************
 * Copyright (c) 2004, 2009 Actuate Corporation.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * https://www.eclipse.org/legal/epl-2.0/.
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 *
 * Contributors:
 *  Actuate Corporation  - initial API and implementation
 *******************************************************************************/
package org.eclipse.birt.data.engine.impl;

import java.util.Collections;
import java.util.Set;
import java.util.TimerTask;
import java.util.concurrent.ConcurrentHashMap;

/**
 *
 */
public class CancelManager extends TimerTask {
	//
	private final Set<ICancellable> cancellables = Collections
			.newSetFromMap(new ConcurrentHashMap<ICancellable, Boolean>());

	/**
	 *
	 * @param cancellable
	 */
	public void register(ICancellable cancellable) {
		cancellables.add(cancellable);
	}

	/**
	 *
	 * @param cancellable
	 */
	public void deregister(ICancellable cancellable) {
		cancellables.remove(cancellable);
	}

	/*
	 * (non-Javadoc)
	 *
	 * @see java.lang.Runnable#run()
	 */
	@Override
	public void run() {
		doCancel();
	}

	public void doCancel() {
		for (ICancellable cancellable : cancellables.toArray(ICancellable[]::new)) {
			if (cancellable.doCancel()) {
				cancellable.cancel();
			}
		}
	}
}
```

Reply:
While you write is true that a concurrent map can be used, it is definitely not bulletproof proof in concurrent environment. With that said i am not sure how this thing works and I opted to use a Map with synchronized. Also a Set uses a Map internally so i am not sure what the benefit will be.

As for the key generation, again i am not sure how this code is supposed to work and i wanted to be sure that it actually works.

Reply:
Adding methods for an ID and generating a UUID just so that you can create what is  effectively a set makes no sense to me and is not minimally invasive either.  A concurrent map is documented that "all operations are thread-safe" so I'm not sure how that is not bullet proof.  Now I see the class has a factory method for creating a set, so this simplification is more direct:
```
/*******************************************************************************
 * Copyright (c) 2004, 2009 Actuate Corporation.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * https://www.eclipse.org/legal/epl-2.0/.
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 *
 * Contributors:
 *  Actuate Corporation  - initial API and implementation
 *******************************************************************************/
package org.eclipse.birt.data.engine.impl;

import java.util.Set;
import java.util.TimerTask;
import java.util.concurrent.ConcurrentHashMap;

/**
 *
 */
public class CancelManager extends TimerTask {
	//
	private final Set<ICancellable> cancellables = ConcurrentHashMap.newKeySet();

	/**
	 *
	 * @param cancellable
	 */
	public void register(ICancellable cancellable) {
		cancellables.add(cancellable);
	}

	/**
	 *
	 * @param cancellable
	 */
	public void deregister(ICancellable cancellable) {
		cancellables.remove(cancellable);
	}

	/*
	 * (non-Javadoc)
	 *
	 * @see java.lang.Runnable#run()
	 */
	@Override
	public void run() {
		doCancel();
	}

	public void doCancel() {
		for (ICancellable cancellable : cancellables.toArray(ICancellable[]::new)) {
			if (cancellable.doCancel()) {
				cancellable.cancel();
			}
		}
	}
}
```
Looking in the Eclipse SDK's sources, I see this specific map implementation is widely used there:

![image](https://github.com/eclipse-birt/birt/assets/208716/46d1fc46-d034-41aa-953a-85b52de5e333)

Reply:
ok, updated it according to suggestion

Reply:
FYI, it's generally better to amend the commit and force push it so that there is a single commit.